xds/internal/balancer/ringhash/picker.go

```diff
@@ -0,0 +1,140 @@
+/*
+ *
+ * Copyright 2021 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package ringhash
+
+import (
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/status"
+)
+
+type picker struct {
+	ring *ring
+}
+
+func newPicker(ring *ring) *picker {
+	return &picker{ring: ring}
+}
+
+// handleRICS generates pick result if the entry is in Ready, IDLE, Connecting
+// or Shutdown. TransientFailure will be handled specifically after this
+// function returns.
+//
+// The first return value indicates if the state is in Ready, IDLE, Connecting
+// or Shutdown. If it's true, the PickResult and error should be returned from
+// Pick() as is.
+func handleRICS(e *ringEntry) (bool, balancer.PickResult, error) {
+	switch state := e.sc.effectiveState(); state {
+	case connectivity.Ready:
+		return true, balancer.PickResult{SubConn: e.sc.sc}, nil
+	case connectivity.Idle:
+		// Trigger Connect() and queue the pick.
+		e.sc.connect()
+		return true, balancer.PickResult{}, balancer.ErrNoSubConnAvailable
+	case connectivity.Connecting:
+		return true, balancer.PickResult{}, balancer.ErrNoSubConnAvailable
+	case connectivity.TransientFailure:
+		// Return ok==false, so TransientFailure will be handled afterwards.
+		return false, balancer.PickResult{}, nil
+	case connectivity.Shutdown:
+		// Shutdown can happen in a race where the old picker is called. A new
+		// picker should already be sent.
+		return true, balancer.PickResult{}, balancer.ErrNoSubConnAvailable
+	default:
+		// Should never reach this. All the connectivity states are already
+		// handled in the cases.
+		return true, balancer.PickResult{}, status.Errorf(codes.Unavailable, "SubConn has undefined connectivity state: %v", state)
```

Comment:
Error log here too so tests detect it.

Reply:
Logger isn't available now (it will be built by the balancer later). I added an TODO for it.